add dashboard modules

imjasonh opened the pull request:
This moves modules in from https://github.com/chainguard-dev/terraform-cloudrun-dashboard -- after this is merged we can archive that repo.

The new directory `./dashboard` includes:
- `service` and `job`, which take a multi-regional Cloud Run Service or Job name and presents common metrics for them
- `topic`, which takes a Pub/Sub topic prefix and presents common metrics for its subscribers, assuming they have that topic name as their subscription name prefix
- `tiles`, which provide reusable dashboard tile definitions, which we can extend over time.

It also defines dashboards for the ingress and recorder services, and broker topics.

mattmoor commented on cloudevent-broker/main.tf:
```diff
@@ -13,3 +13,9 @@ resource "google_pubsub_topic" "this" {
   message_retention_duration = "600s"
 }
+
+module "dashboard" {
```
This collides with the invocation in `ingress.tf`, but I'd like to see if we can combine these into a unified broker dashboard using your building blocks 🤩

mattmoor commented on dashboard/job/variables.tf:
```diff
@@ -0,0 +1,9 @@
+variable "project_id" { type = string }
+variable "job_name" { type = string }
+
+variable "alert_policies" {
+  type = map(object({
+    id = string
+  }))
+  default = {}
```
I've been trying to add detailed comments to the `variables` and `outputs` so our auto-docs are 💯

mattmoor replied:
I've also been trying to add fairly decent usage comments to the `README.md` files.

imjasonh replied:
Can I give you an IOU for the alert policy stuff for now? That's my next chunk of work and I might completely redo how it's structured.

mattmoor commented on dashboard/tiles/latency/main.tf:
```diff
@@ -0,0 +1,39 @@
+variable "title" { type = string }
+variable "group_by_fields" { default = [] }
+variable "filter" { type = list(string) }
```
nit: split `variables.tf` and `outputs.tf` 🙏

mattmoor replied:
oh fun, the whole thing is inputs/outputs 🤔

imjasonh replied:
Yeah, these tiles are really simple little things. I could split them up but it's just more tab-flipping when you iterate on them.

mattmoor commented on dashboard/topic/dashboard.tf:
```diff
@@ -0,0 +1,76 @@
+  filter = [
+    "resource.type=\"pubsub_subscription\"",
+    "metric.type=\"pubsub.googleapis.com/subscription/push_request_latencies\"",
+    "resource.label.\"subscription_id\"=monitoring.regex.full_match(\"${var.topic_prefix}.*\")",
```
`topic_prefix` seems wrong.  I think you want `subscription_prefix`?

mattmoor commented on .github/workflows/documentation.yaml:
```diff
@@ -15,6 +15,9 @@ jobs:
         - cloudevent-trigger
         - cloudevent-recorder
         - networking
+        - dashboard/service
+        - dashboard/job
+        - dashboard/topic
```
Do we want the tiles too?

mattmoor replied:
subscription?

imjasonh replied:
Ehhh, maybe. They're so simple I think we'd end up having most of the content of the directories be comments, and generated markdown from comments.

Which maybe isn't a bad thing! I just don't want to have to write it all right now 😆

mattmoor replied:
LOL, ok.  I've been trying to keep the bar here relatively high, so let's take the AI to chase this once we feel like the shape is something we're really bought into?

imjasonh replied:
re: topic vs subscription, wdyt about `dashboard/subscriptions` (with an `s`)?

The PubSub dashboard is weird because it doesn't just go by the (regionally-unique) name of the thing, it goes by prefixes on (globally-unique) subscription names.

imjasonh replied:
The sent-events tile for instance relies on the fact that `subscription_prefix` is also a topic name.

mattmoor replied:
We should be able to filter subscriptions metrics on `topic_id` without assuming the subscription prefix matches.  I was playing with this in metrics explorer earlier.

mattmoor replied:
But let's land the widgets / cloudrun dashboards, and then talk pubsub next

imjasonh commented:
Dropped alert policies for now, I'll readd it later. Also dropped the project_id variable, which was only used to power the url output, which I've also removed.

mattmoor commented on cloudevent-broker/dashboard.tf:
```diff
@@ -0,0 +1,88 @@
+  title  = "Events Sent"
+  filter = [
+    "resource.type=\"pubsub_topic\"",
+    "metric.type=\"pubsub.googleapis.com/topic/send_request_count\"",
```
We should filter this on the topic name prefix

mattmoor replied:
See my other comment, but I think you factored this out (with prefix!) and then forgot to clean this up?

mattmoor replied:
oh weird, no.

mattmoor replied:
The "topic" dashboard shows "subscription" stats 🤔

mattmoor commented on cloudevent-broker/main.tf:
```diff
@@ -13,3 +13,9 @@ resource "google_pubsub_topic" "this" {
   message_retention_duration = "600s"
 }
+
+module "topic-dashboard" {
+  source       = "../dashboard/topic"
+  project_id   = var.project_id
+  topic_prefix = var.name
+}
```
How is this different from `dashboard.tf`? 🤔

imjasonh replied:
Oh it's not! This was an artifact of refactoring. Removed.

mattmoor commented on dashboard/tiles/alert/main.tf:
```diff
@@ -0,0 +1,10 @@
+variable "alert_name" { type = string }
+
+// https://cloud.google.com/monitoring/api/ref_v3/rest/v1/projects.dashboards#AlertChart
+output "tile" {
```
These are actually widgets

mattmoor commented on dashboard/topic/dashboard.tf:
drop for now

mattmoor commented on dashboard/subscription/dashboard.tf:
drop for now

mattmoor commented on dashboard/subscription/README.md:
drop for now

mattmoor commented on .github/workflows/documentation.yaml:
```diff
@@ -15,6 +15,9 @@ jobs:
         - networking
+        - dashboard/service
+        - dashboard/job
+        - dashboard/topic
```
drop for now
```suggestion
```

mattmoor commented on dashboard/widgets/xy/main.tf:
```diff
@@ -0,0 +1,46 @@
+variable "secondary_reduce" { default = "REDUCE_NONE" }
+
+// https://cloud.google.com/monitoring/api/ref_v3/rest/v1/projects.dashboards#XyChart
+output "tile" {
```
We should rename the outputs too.
